[WIP] Connect Account tokenization

danj-stripe opened the pull request:
## Summary

Simple summary of what was changed.

## Motivation

Why are you making this change? If it's for fixing a bug, if possible, please include a code snippet or example project that demonstrates the issue.

## Testing

How was the code tested? Be as specific as possible.

danj-stripe commented on Stripe/PublicHeaders/STPAPIClient.h:
```diff
@@ -127,6 +127,19 @@ static NSString *const STPSDKVersion = @"11.5.0";
+ */
+@interface STPAPIClient (ConnectAccounts)
+
+- (void)createTokenWithConnectAccount:(id)obj completion:(__nullable STPTokenCompletionBlock)completion;
```
obj needs to turn into:
```
account: {
  legal_entity: {...},
  tos_shown_and_accepted: true,
}
```
with either `legal_entity` **or** `tos_shown_and_accepted`, or both.

I don't know whether I want to create another `...Params` class, or change this method to take two nullable parameters.

I also think it's unfortunate that `Bool?` is `nullable NSNumber` in Obj-C

bg-stripe replied:
Hmm, I think it'd be better to make this method take parameters rather than make a parent `AccountParams` class, but don't feel strongly.

Is it necessary to model `tos_shown_and_accepted` as a `Bool?`, or can we treat `tos_shown_and_accepted: nil` as `tos_shown_and_accepted: false`?

danj-stripe replied:
I don't think that would work. Valid values for `tos_shown_and_accepted` are `{true, nil}`. Some basic use cases:

* `account: { tos_shown_and_accepted: true }`: minimum required to set up a new account
* `account: { legal_entity: { ... }, tos_shown_and_accepted: true }`: setting up new account with some data about the legal entity
* `account: { legal_entity: { ... } }`: valid for updating a pre-existing account
* `account: { tos_shown_and_accepted: false, ... }`: rejected by the server as invalid

As I understand it, the initial purpose for this API is collecting positive affirmation that the user has seen & accepted the ToS, instead of the old world where the platform is allowed to vouch that fact as true. However, it's *also* just as usable for updating the legal entity info directly.


I'll think about `AccountParams` a little more. I like that it's easier to add new fields in the future, and the possibility of providing >1 constructors to help understand the `legal_entity/tos_shown_and_accepted` distinction. But it's a lot just for two optional fields

bg-stripe replied:
ah interesting, if that's the case then perhaps an `AccountParams` object with multiple constructors makes sense. thanks for the explanation!

danj-stripe commented on Stripe/STPLegalEntityParams.h:
```diff
@@ -0,0 +1,48 @@
+};
+
+
+@interface STPLegalEntityParams : STPPersonParams
```
Any problems foreseen by inheriting from `STPPersonParams`, which conforms to `STPFormEncodable`?

bg-stripe replied:
hmm, I can't think of any

danj-stripe commented on Stripe/STPLegalEntityParams.h:
```diff
@@ -0,0 +1,48 @@
+};
+
+
+@interface STPLegalEntityParams : STPPersonParams
```
Do `STPVerificationParams` and `STPPersonParams` belong in their own files? Or are they fine here?

They're *only* used here (for now?), **and** the inheritance also links `STPLegalEntityParams` with `STPPersonParams` pretty closely.

bg-stripe replied:
I think they're fine together!

danj-stripe commented:
@bg-stripe WDYT? Does this look ~reasonable?

bg-stripe commented:
this looks reasonable to me!

danj-stripe commented:
This includes the commit for #881, and I think it's now complete.

r? @bg-stripe

bg-stripe commented on Stripe/STPAPIClient.m:
```diff
@@ -288,6 +288,19 @@ - (void)createTokenWithPersonalIDNumber:(NSString *)pii completion:(__nullable S
+
+@implementation STPAPIClient (ConnectAccounts)
+
+- (void)createTokenWithConnectAccount:(__unused id)account completion:(__nullable STPTokenCompletionBlock)completion {
```
replace `__unused id` with `STPConnectAccountParams *``

bg-stripe commented on Stripe/STPAPIClient.m:
```diff
@@ -288,6 +288,19 @@ - (void)createTokenWithPersonalIDNumber:(NSString *)pii completion:(__nullable S
+
+- (void)createTokenWithConnectAccount:(__unused id)account completion:(__nullable STPTokenCompletionBlock)completion {
+    NSMutableDictionary *params = [[STPFormEncoder dictionaryForObject:account] mutableCopy];
+    [[STPTelemetryClient sharedInstance] addTelemetryFieldsToParams:params];
```
I think we probably don't need to add telemetry data.

danj-stripe replied:
Papertrail: bg and I discussed in Slack. Android sends telemetry for all tokens. We think this is fine, but I have an open ask out to the team that owns telemetry.

bg-stripe commented on Stripe/STPAddress.m:
```diff
@@ -42,14 +44,8 @@ + (NSDictionary *)shippingInfoForChargeWithAddress:(nullable STPAddress *)addres
-    addressDict[@"country"] = address.country;
-    params[@"address"] = [addressDict copy];
+    // Re-use STPFormEncoder
+    params[@"address"] = [STPFormEncoder dictionaryForObject:address];
```
👍

bg-stripe commented on Stripe/STPLegalEntityParams.m:
```diff
@@ -0,0 +1,150 @@
+    return @{};
+}
+
+- (void)setAdditionalAPIParameters:(__unused NSDictionary *)additionalAPIParameters {
```
👍

bg-stripe commented on Stripe/STPConnectAccountParams.m:
```diff
@@ -0,0 +1,65 @@
+//
+
+#import "STPConnectAccountParams.h"
+
```
NS_ASSUM_NONNULL

bg-stripe commented on Stripe/STPLegalEntityParams.m:
```diff
@@ -0,0 +1,150 @@
+//  Copyright © 2017 Stripe, Inc. All rights reserved.
+//
+
+#import "STPLegalEntityParams.h"
```
NS_ASSUME_NONNULL

danj-stripe commented:
Also, cc @ksun-stripe, who implemented this in stripe/stripe-android#492

danj-stripe commented:
@bg-stripe I've addressed your comments (thanks for all the help!), and fixed some other problems. Travis has a 1000+ build backlog and is running at half capacity. If you want to re-review now you can, or I'll assign to you again if/when the build has succeeded.
